### DISCUSSION TITLE
What is the difference between all the BPM plugins? Which one is the best?

### DISCUSSION THREAD

**Discussion by shimizurei** [beetbox/beets#5208] [Q&A] [answered]
I just want an accurate BPM for the BPM field.

Looking at the docs there is:

- https://github.com/adamjakab/BeetsPluginBpmAnalyser
- https://beets.readthedocs.io/en/stable/plugins/bpm.html
- https://beets.readthedocs.io/en/latest/plugins/autobpm.html

I'm working on my config file (haven't used Beets in years) and I have no idea which one to use. I just know I won't be using the keystroke one because I'd mess that up.

**Comment by adamjakab** [comment, marked as answer]
Hi @shimizurei,

The second one `bpm`: "lets you to get the tempo (beats per minute) of a song by tapping out the beat on your keyboard." . Even though I think it is a good exercise probably a time consuming choice if you have Gigs of music to go through.

I am the author of the first one `BeetsPluginBpmAnalyser`  (so I might be biased ;) ) and I believe that at that time the `autobpm` plugin did not exist. I think that the main difference is in the music analysis library they use: [aubio](https://pypi.org/project/aubio/) and [librosa](https://github.com/librosa/librosa/). I do not know if one library is more performant than the other and/or if the calculated values are more precise with aubio or with librosa.

I also see some slight differences in the configuration options where `BeetsPluginBpmAnalyser` allow for a more fine-tuned configuration:
```
bpmanalyser:
  auto: no
  dry-run: no
  write: yes
  threads: 2
  force: no
  quiet: no
```

Maybe you could install both and run a comparison. That could be interesting.

**EDIT**: Pls note that (probably due to some changes in the aubio library) atm the `BeetsPluginBpmAnalyser` plugin fails with error; `RuntimeError: AUBIO ERROR: source_wavread: Failed opening /mnt/Music/xyz.mp3 (could not find RIFF header)`. 
I just found this thread: https://github.com/aubio/aubio/issues/116 which reports the same issue.
I will look into it and get it fixed.

**EDIT2**: I fixed the plugin. I now uses the [pydub](https://github.com/jiaaro/pydub) library (which relies on ffmpeg or libav) to convert the audio files to wav before analyzing them with aubio.

I hope this helps.
